geoAssembler/optimiser/centre.py

```diff
@@ -0,0 +1,180 @@
+
+        centre_offset: pair of values
+            Sets the centre position of the search grid.
+
+        pool: a multiprocessing Pool object
+            Used to enable multiprocessing over multiple threads, it is
+            recommended to only use 32 threads, the number of threads
+            should increase if larger grids are used.
+
+        verbose: Bool
+            Set to true to print status and progress messages.
+        """
+        res_tuple = namedtuple("FindCentreResult", "centre array xs ys")
+        xs = np.arange(-radius, radius+stepsize, stepsize) + centre_offset[0]
+        ys = np.arange(-radius, radius+stepsize, stepsize) + centre_offset[1]
+
+        if verbose:
+            print(f"Trying {len(xs)*len(ys)} combinations, "
+                  f"radius {radius}, stepsize {stepsize} - ", end='')
+
+        if pool is not None:
+            min_array = pool.map(self._loss_function, product(xs, ys))
+        else:
+            min_array = map(self._loss_function, product(xs, ys))
+
+        min_array = np.array(list(min_array))
+        min_array = np.reshape(min_array, (len(xs), len(ys)))
+
+        centre_idx = [x[0] for x in np.where(min_array == np.amin(min_array))]
+        centre = np.array([xs[centre_idx[0]], ys[centre_idx[1]]])
+
+        if verbose:
+            print(f"found centre at -> {centre}")
+
+        res = res_tuple(
+            centre,
+            min_array,
+            xs, ys
+        )
+
+        return res
+
+    def optimise(self, r_step_pairs=[(50, 10), (10, 2), (3, 0.5)],
+                 centre_offset=[0, 0], pool=None, verbose=True):
+        """
+        Function which performs a grid-search which goes from coarse to fine
+        coordinates, defined by `r_step_pairs`.
+
+        Parameters
+        ----------
+
+        r_step_pairs: list of tuples
+            A list of tuple pair of radius and step size, where the optimum
+            coordinates of the previous grid are used as the centre for the
+            subsequent grid.
+
+            e.g. `[(50, 10), (10, 2)]` creates a grid of pixel of pixel
+            coordinates from -50 to +50, with steps of 10, so an 11x11 grid.
+
+            The optimal position is then fed into the next step, which uses a
+            grid from -10 to +10, with steps of 2, centred on the previous pos.
+
+        pool: a multiprocessing Pool object
+            Used to enable multiprocessing over multiple threads, it is
+            recommended to only use 32 threads, the number of threads
+            should increase if larger grids are used.
+        """
+        res_tuple = namedtuple(
+            "OptimiseResult", "optimal_quad_positions optimal_offset results"
+        )
+
+        results = []
+        for radius, stepsize in r_step_pairs:
+            res = self._find_centre(
+                radius, stepsize,
+                centre_offset,
+                verbose=verbose,
+                pool=pool
+            )
```

Comment:
I'm still in two minds about the 'zoom in and enhance' custom optimiser here. It does mean we can understand exactly what the code is doing, and easily make plots to inspect the intermediate steps. And it seems to work, which is never something to sneeze at!

On the other hand, numerical optimisation is a well established thing, and there are many battle-tested algorithms for doing it. [scipy.optimize](https://docs.scipy.org/doc/scipy/reference/optimize.html) has several, and there's also [scikit-optimize](https://scikit-optimize.github.io/stable/), which includes [visualisation tools](https://scikit-optimize.github.io/stable/modules/plots.html) for inspecting what's happening. Doing something smarter than grid search might mean it's easy to extend this to find quadrant positions (i.e. fitting 8 variables).

Can I ask you to investigate what it would look like to use an off-the-shelf optimiser? Is it easy to fit it into the code, and what do we lose/gain in inspecting the results? I might start with scikit-optimize, which looks more approachable than scipy. If you don't have time for this, I don't mind landing the feature with the custom optimiser, but I'm keeping my fingers crossed that it would be an easy win.

Reply:
Yeah I looked into this a bit and just wrote my own as it was easier than figuring out how to use the existing libraries.

Far as I remember when I originally did this, none of the scipy optimisers work for this as they're designed for continuous/floating value optimisation, and the steps/gradients for that are hard coded into the optimisers. scikit optimize does have a way to define an [integer search space](https://scikit-optimize.github.io/stable/modules/generated/skopt.space.space.Integer.html), but that does (far as I could tell from the docs) the same thing mine does, then I [read the the "docs" ](https://scikit-optimize.github.io/stable/modules/space.html#space) and given the info there is seemed easier to make my own.

Reply:
> one of the scipy optimisers work for this as they're designed for continuous/floating value optimisation

I'm a bit puzzled - isn't the centre location a continuous variable? I'd expect pyFAI to be able to use greater than pixel accuracy for this. Not that we necessarily need that precision, given that we're 'snapping' pixels to a single grid anyway, but it should mean we're optimising in a continuous space.

Reply:
Ah yes I didn't explain that well, I still wanted to restrict it to integer space at the start and to have it only look at sub-pixel regions later, since there I imagined an optimiser might get stuck looking at the sub-pixel positions too early. Now that I actually looked into it it's actually very easy to implement, turns out I was looking in the wrong part of the docs :P

Summary of my adventures would be that all but one of the optimisation methods did not work well or at all for this problem. In the end the differential evolution method of global optimisation from scipy finished in a vaguely sensible amount of time, actually gave good results, **and** has multiprocessing via workers built in, so I'll swap over to using that. 

Thanks for prompting me to look into this more, I originally tried only the basic optimiser, saw it didn't work, and wrote them off. Turns out the penultimate method in the docs is the one that works lol.

--------------------------------

This is a longer summary, mostly for myself for future reference, so feel free to skip it.

I tried `gp_minimize` but everything I tried ended up being both wwaayy slower, and the results were worse as well. The Bayesian optimisation it uses really likes getting stuck at the edges of the bounds, even though the minimum isn't near the edges, I really don't get why it's behaving like that, spent a while trying to fix it and gave up. It also has the (to me) very odd behaviour of repeatedly testing the same argument, so it doesn't seem to cache results or be aware of what regions it has already gone over.

Also tried a few of the scipy optimize minimisation methods, went through most of the unconstrained and constrained multivariate function minimizers, and none of them worked very well out of the box either.

That was... vaguely predictable I guess, the function that's being minimised isn't the best and jumps around a fair bit, so I moved on to trying to use global optimisers.

SHGO was fast, but didn't find the minimum. Dual annealing was slow to the point where I gave up running it, and there is no simple way to parallelise it; same with basin hopping.

I also found that scipy optimize has a 'brute' method, which does exactly what I do with my grid search method.

Mercifully in the end the differential evolution method  finished in a vaguely sensible amount of time, actually gave good results, **and** has multiprocessing via workers built in, so I'll swap over to using that.